Rebase Playlist core

sangwoo108 opened the pull request:
According to the original commit log, the core

* manages playlist. Playlist core API
* is exported via PlaylistService.
* downloads thumbnail and media files

During rebasing, addressed comments on original PR
* https://github.com/brave/brave-core/pull/5733/files#r612595759
* https://github.com/brave/brave-core/pull/5733/files#r612607128

<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/23622

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
PlaylistBrowserTest.*

simonhong commented on components/playlist/playlist_media_file_downloader.h:
```diff
@@ -0,0 +1,127 @@
+
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+
+  SimpleURLLoaderList url_loaders_;
```
Can we use `api_request_helper::APIRequestHelper` in this PR instead of using `SimpleURLLoader` directly?

sangwoo108 replied:
Cool! I made a more API to APIRequestHelper for it.

simonhong replied:
Wow, cool! 👍🏼

simonhong commented on components/playlist/playlist_service.h:
```diff
@@ -0,0 +1,161 @@
+  void RemovePlaylist(const std::string& id);
+
+  // index.html is only used for demo (brave://playlist)
+  void GenerateIndexHTMLFile(const base::FilePath& playlist_path);
```
I think we don't need to generate this anymore.
Each playlist item has its own html file for playing. It was only for demo purpose.
In production, new webui will have whole ui for managing/playing playlist items.

sangwoo108 replied:
Yep . Done.

simonhong reviewed (approved):
++

simonhong commented on components/playlist/features.cc:
```diff
@@ -0,0 +1,19 @@
+namespace playlist {
+namespace features {
+
+const base::Feature kPlaylist{"PlaylistName",
```
nit: maybe `Playlist` ? I assume `PlaylistName` was just temporary string before getting final name.

sangwoo108 replied:
done

simonhong commented on components/playlist/buildflags/buildflags.gni:
```diff
@@ -0,0 +1,10 @@
+import("//extensions/buildflags/buildflags.gni")
+
+declare_args() {
+  enable_playlist = enable_extensions || is_android
```
Hmm, I think `is_win || is_mac || is_linux || is_android` would be correct.
As we'll not use extension for playlist, `enable_extensions` is not needed?

sangwoo108 replied:
Yep. Done.

simonhong commented on components/playlist/playlist_data_source.h:
```diff
@@ -0,0 +1,56 @@
+  bool AllowCaching() override;
+
+ private:
+  PlaylistService* service_;
```
Let's apply `raw_ptr` in this PR!

sangwoo108 replied:
Done.

simonhong reviewed (commented):
Left some trivial comments.

sangwoo108 commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,339 @@
+  delegate_->OnMediaFileReady(id, media_file_path_key, media_file_path);
+}
+
+void PlaylistMediaFileDownloader::GenerateSingleMediaFile(
```
It seems videos and audios in a playlist are merged into a single file. if so, this will make it hard:
* to edit playlist, e.g. remove an item from the list.
* share same file across playlists

What can be benefit over these limits?

simonhong replied:
Single file per each playlist item not whole playlist.
IIRC, youtube gives multiple sources for one clip. So, this method was introduced.

sangwoo108 replied:
Aha, yes. youtubedown.js seems to suggest multiple source based on resolutions and others. then, Maybe we can remove this as our script gives us one url.

simonhong commented:
If you'll revisit component part(youtubedown.js or iOS's script) later, how about making that part abstract instead of shipping youtubedown.js?

sangwoo108 commented:
> If you'll revisit component part(youtubedown.js or iOS's script) later, how about making that part abstract instead of shipping youtubedown.js?

Yes. That would be better. Renamed youtube_down_xxx -> media_detector_xxx and removed registration part.

goodov commented on components/playlist/playlist_service.cc:
```diff
@@ -0,0 +1,399 @@
+base::SequencedTaskRunner* PlaylistService::task_runner() {
+  if (!task_runner_) {
+    task_runner_ = base::ThreadPool::CreateSequencedTaskRunner(
+        {base::MayBlock(), base::TaskPriority::BEST_EFFORT,
```
same. `USER_VISIBLE` is more appropriate here.

goodov commented on components/playlist/playlist_service.cc:
```diff
@@ -0,0 +1,399 @@
+const base::FilePath::StringType kThumbnailFileName(
+    FILE_PATH_LITERAL("thumbnail"));
+
+void DeleteDir(const base::FilePath& path) {
```
just use `base::GetDeletePathRecursivelyCallback`.

goodov commented on components/playlist/playlist_service.cc:
```diff
@@ -0,0 +1,399 @@
+PlaylistService::~PlaylistService() = default;
+
+void PlaylistService::Shutdown() {
+  download_request_manager_.reset();
```
reset other members that still can do something?

goodov commented on components/playlist/playlist_service.cc:
```diff
@@ -0,0 +1,399 @@
+  }
+
+  task_runner()->PostTaskAndReplyWithResult(
+      FROM_HERE, base::BindOnce(&GetOrphanedPaths, base_dir_, ids),
```
std::move(ids)

goodov commented on components/playlist/playlist_service_observer.h:
```diff
@@ -0,0 +1,26 @@
+      const PlaylistChangeParams& params) = 0;
+
+ protected:
+  ~PlaylistServiceObserver() override {}
```
no need to have this.

goodov commented on components/playlist/playlist_types.cc:
```diff
@@ -0,0 +1,70 @@
+
+PlaylistInfo::PlaylistInfo() = default;
+PlaylistInfo::~PlaylistInfo() = default;
+PlaylistInfo::PlaylistInfo(const PlaylistInfo& rhs) {
```
shouldn't this just be `= default`?
also if a class declares a copy-constructor, it should declare an assignment operator.

To be absolutely perfect, we also need to have rvalue noexcept constructor/operator=.
https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types

sangwoo108 commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,339 @@
+      break;
+    }
+
+    if (!AppendToFileThenDeleteSource(media_file_source_path,
```
I believe so. To be honest, we can't test the original code which spits out separate audio and video file for now. But it seems to have worked. 

@bsclifton @simonhong Could you tell us about this more?

simonhong replied:
Well.. I can't remember whether this final single media file is generated from multiple splitted source files
or just one source file. Not sure but maybe data type is list but only one soure file url was there.
@pilgrim-brave It's been a long time, but can you remember?

simonhong replied:
I think this detail comes from youtubedown.js. If our new scripts give one source url, I think we don't need this method.

sangwoo108 replied:
> If our new scripts give one source url, I think we don't need this method.

Yes. That's what I'm hoping.

sangwoo108 commented on components/playlist/playlist_media_file_download_manager.cc:
```diff
@@ -0,0 +1,152 @@
+        std::move(pending_media_file_creation_jobs_.front()));
+    pending_media_file_creation_jobs_.pop();
+    const std::string playlist_id =
+        *playlist_item.FindStringKey(kPlaylistIDKey);
```
Yes. Added DCHECK.

sangwoo108 commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,339 @@
+base::SequencedTaskRunner* PlaylistMediaFileDownloader::task_runner() {
+  if (!task_runner_) {
+    task_runner_ = base::ThreadPool::CreateSequencedTaskRunner(
+        {base::MayBlock(), base::TaskPriority::BEST_EFFORT,
```
Thanks! Done.

sangwoo108 commented on components/playlist/playlist_download_request_manager.cc:
```diff
@@ -0,0 +1,214 @@
+  DCHECK_GE(in_progress_youtube_urls_count_, 0);
+  in_progress_youtube_urls_count_++;
+
+  LOG(ERROR) << __FUNCTION__ << " " << media_detector_script_;
```
oops, removed.

sangwoo108 commented on browser/playlist/test/playlist_browsertest.cc:
```diff
@@ -0,0 +1,358 @@
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    // HTTPS server only serves a valid cert for localhost, so this is needed
+    // to load pages from other hosts without an error
+    command_line->AppendSwitch(switches::kIgnoreCertificateErrors);
```
Didn't know that. Thanks for the pointer. Done.

sangwoo108 commented on components/playlist/playlist_download_request_manager.cc:
```diff
@@ -0,0 +1,213 @@
+  const bool has_audio = list.size() > 1;
+
+  // Get clean playlist name.
+  params.playlist_name = *list[0].FindStringKey("file");
```
You're right. Added check here.

sangwoo108 commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,339 @@
+                                  const base::FilePath& destination_file_path,
+                                  char* read_buffer,
+                                  size_t read_buffer_size) {
+  base::ScopedFILE source_file(base::OpenFile(source_path, "rb"));
```
Yeah. Also base::File provides seems to provide us better API! Thanks. done.

sangwoo108 commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,341 @@
+
+  base::File dest_file(destination_file_path,
+                       base::File::FLAG_OPEN | base::File::FLAG_APPEND);
+  if (dest_file.IsValid())
```
```suggestion
  if (!dest_file.IsValid())
```

goodov commented on components/playlist/playlist_types.h:
```diff
@@ -0,0 +1,80 @@
+  PlaylistInfo(const PlaylistInfo& rhs);
+  PlaylistInfo& operator=(const PlaylistInfo& rhs);
+  PlaylistInfo(PlaylistInfo&& rhs);
+  PlaylistInfo& operator=(PlaylistInfo&& rhs);
```
sorry for being too picky, but move constructors/operators should be marked `noexcept`. even though we have exceptions disabled, this is still required.

https://google.github.io/styleguide/cppguide.html#noexcept
> Specifying move constructors as noexcept improves performance in some cases, e.g., std::vector<T>::resize() moves rather than copies the objects if T's move constructor is noexcept

sangwoo108 replied:
> sorry for being too picky, but move constructors/operators should be marked noexcept. even though we have exceptions disabled, this is still required.

No worries. I didn't know that there's performance improvement with `noexcept`. Thanks!

goodov commented on components/playlist/playlist_service.h:
```diff
@@ -0,0 +1,158 @@
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+  raw_ptr<PrefService> prefs_ = nullptr;
+
+  base::WeakPtrFactory<PlaylistService> weak_factory_;
```
{this}

goodov commented on components/playlist/playlist_media_file_downloader.h:
```diff
@@ -0,0 +1,128 @@
+
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+
+  base::WeakPtrFactory<PlaylistMediaFileDownloader> weak_factory_;
```
{this}

goodov commented on components/playlist/playlist_media_file_downloader.cc:
```diff
@@ -0,0 +1,341 @@
+      base::File::FLAG_CREATE | base::File::FLAG_WRITE);
+  if (!unified_media_file.IsValid())
+    return false;
+  unified_media_file.Close();
```
keep it opened and just pass it into `AppendToFileThenDeleteSource` as a ptr.

goodov commented on components/playlist/playlist_data_source.cc:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  scoped_refptr<base::RefCountedMemory> bytes;
+  bytes = base::MakeRefCounted<base::RefCountedBytes>(
```
pls create this right in the `ReadFileToString` (maybe rename to `ReadFileToRefCountedBytes`).
no need to copy the byte array on UI thread.

goodov commented on components/playlist/media_detector_component_installer.cc:
```diff
@@ -0,0 +1,149 @@
+  return update_client::InstallerAttributes();
+}
+
+[[maybe_unused]] void OnRegistered(const std::string& component_id) {
```
is this a WIP function? it will be used in a next PR or smth?

sangwoo108 replied:
Yes. But it looks like removing it from this PR better.

sangwoo108 commented:
@goodov Anything else?

goodov commented on browser/playlist/test/playlist_browsertest.cc:
```diff
@@ -0,0 +1,372 @@
+    host_resolver()->AddRule("*", "127.0.0.1");
+
+    // Set up embedded test server to handle fake responses.
+    https_server_.reset(new net::EmbeddedTestServer(
```
make_unique

goodov commented on browser/sources.gni:
```diff
@@ -437,6 +439,7 @@ brave_chrome_browser_deps += brave_browser_importer_deps
 brave_chrome_browser_deps += brave_browser_ipfs_deps
 brave_chrome_browser_deps += brave_browser_new_tab_deps
 brave_chrome_browser_deps += brave_browser_notifications_deps
+brave_chrome_browser_deps += brave_browser_playlist_deps
```
should go after `brave_browser_permissions_deps`

goodov commented on components/playlist/playlist_data_source.h:
```diff
@@ -0,0 +1,56 @@
+  bool AllowCaching() override;
+
+ private:
+  raw_ptr<PlaylistService> service_;
```
vars should go after methods.

goodov commented on components/playlist/playlist_download_request_manager.cc:
```diff
@@ -0,0 +1,204 @@
+  // Never allow running in main world (0).
+  DCHECK(id > content::ISOLATED_WORLD_ID_GLOBAL);
+  // Only allow ID to be set once.
+  DCHECK(g_playlist_javascript_world_id == kInvalidWorldID);
```
`DCHECK(!PlaylistJavaScriptWorldIdIsSet());`

goodov commented on components/playlist/playlist_download_request_manager.cc:
```diff
@@ -0,0 +1,204 @@
+void PlaylistDownloadRequestManager::SetPlaylistJavaScriptWorldId(
+    const int32_t id) {
+  // Never allow running in main world (0).
+  DCHECK(id > content::ISOLATED_WORLD_ID_GLOBAL);
```
`id > ISOLATED_WORLD_ID_CONTENT_END`

goodov commented on components/playlist/playlist_download_request_manager.h:
```diff
@@ -0,0 +1,112 @@
+  std::list<std::string> pending_youtube_urls_;
+
+  // Used to inject youtubedown.js to get playlist item metadata to download
+  // its media files/thumbnail mages and get titile.
```
`images and get title`

goodov reviewed (approved):
lgtm with some nits.

It would also be great to have unit tests for downloaders and other parts that behave like state machines. I don't know what and where, but some of them might break/crash the browser in some situations.

There's also a lot of `base::Value` accessors that dereference without checks, not sure if they'll break at some point or these places are well protected with some upstream calls.

sangwoo108 commented:
> It would also be great to have unit tests for downloaders and other parts

Yes. I agree. Let me do that when I migrate the new script.

> base::Value accessors that dereference without checks,

Added more checks for base::Value. 

---

The android test failures seems to be unrelated to this PR so let me merge this. Thank you all for reviewing this!

sangwoo108 commented:
Reportedly, the android failures above keeps happening. I think I'll have to disable component/playlist on Android.
